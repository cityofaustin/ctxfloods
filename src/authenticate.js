const authenticate = function(lokka, userType, email, password) {
  describe('When signing in as a ' + userType, () => {
    describe('with the correct username and password', () => {
      it('should generate a valid jwtToken', async () => {
        const response = await lokka.send(`
          mutation($email:String!, $password:String!) {
            authenticate(input: {email: $email, password: $password}) {
              jwtToken
            }
          }
        `,
        {
          email: email,
          password: password
        });

        expect(response.authenticate.jwtToken).not.toBeNull();
      });
    });

    describe('with an incorrect username and password', () => {
      it('should generate a null jwtToken', async () => {
        const response = await lokka.send(`
          mutation($email:String!, $password:String!) {
            authenticate(input: {email: $email, password: $password}) {
              jwtToken
            }
          }
        `,
        {
          email: email,
          password: "wrong"
        });

        expect(response.authenticate.jwtToken).toBeNull();
      });
    });
  });
}

export default authenticate;
