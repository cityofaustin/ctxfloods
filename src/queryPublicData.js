// TODO: Figure out how to use this in the community editor/admin/super admin tests
const queryPublicData = function(lokka) {
  describe('When querying public data', () => {

    it('should get all users', async () => {
      const response = await lokka.send(`
      {
        allUsers {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all communities', async () => {
      const response = await lokka.send(`
      {
        allCommunities {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all statuses', async () => {
      const response = await lokka.send(`
      {
        allStatuses {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status updates', async () => {
      const response = await lokka.send(`
      {
        allStatusUpdates {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status reasons', async () => {
      const response = await lokka.send(`
      {
        allStatusReasons {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status durations', async () => {
      const response = await lokka.send(`
      {
        allStatusDurations {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all crossings', async () => {
      const response = await lokka.send(`
      {
        allCrossings {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all community crossings', async () => {
      const response = await lokka.send(`
      {
        allCommunityCrossings {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get the latest statuses correctly', async () => {
      const response = await lokka.send(`
        {
          allCrossings {
            nodes {
              latestStatus {
                id
              }
            }
          }
        }
      `);

      expect(response).not.toBeNull();
    });

  });
}

export default queryPublicData;
