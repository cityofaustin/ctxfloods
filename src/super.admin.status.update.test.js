import HttpTransport from 'lokka-transport-http';
import Lokka from 'lokka';
import queryPublicData from './queryPublicData';
import authenticate from './authenticate';

const anonLokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql')});
const superAdminEmail = 'superadmin@flo.ods';
const superAdminPassword = 'texasfloods';

async function getToken(email, password) {
  const response = await anonLokka.send(`
    mutation($email:String!, $password:String!) {
      authenticate(input: {email: $email, password: $password}) {
        jwtToken
      }
    }
  `,
  {
    email: email,
    password: password
  });

  return response.authenticate.jwtToken;
}

async function shouldWork(username, password, status, crossing, notes, reason, duration) {
  var originalStatusId;
  var newStatusId;
  var lokka;

  beforeAll(async (done) => {
    getToken(username, password).then((token) => {
      const headers = {
        'Authorization': 'Bearer '+ token
      };
      lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
      done();
    });
  });

  it('before updating it should get the current status', async () => {
    const response = await lokka.send(`
      query($crossing:Int!) {
        crossingById(id:$crossing) {
          latestStatus {
            id
          }
        }
      }
    `,
    {
      crossing: crossing
    });

    originalStatusId = response.crossingById.latestStatus.id;
    expect(response).not.toBeNull();
  });

  it('should update the status', async () => {
    const response = await lokka.send(`
      mutation($status:Int,$crossing:Int,$notes:String,$reason:Int,$duration:Int) {
        newStatusUpdate(input: {
          statusId: $status,
          crossingId: $crossing,
          notes: $notes,
          statusReasonId: $reason,
          statusDurationId: $duration
        }) {
          statusUpdate {
            id
            notes
          }
        }
      }
    `,
    {
      status: status,
      crossing: crossing,
      notes: notes,
      reason: reason,
      duration: duration
    });

    newStatusId = response.newStatusUpdate.statusUpdate.id;
    expect(response.newStatusUpdate.statusUpdate.notes).toMatchSnapshot();
  });

  it('the new status should show up in the DB', async () => {
    const response = await lokka.send(`
      query($crossing:Int!) {
        crossingById(id:$crossing) {
          latestStatus {
            id
            notes
          }
        }
      }
    `,
    {
      crossing: crossing
    });

    expect(response.crossingById.latestStatus.id).not.toEqual(originalStatusId);
    expect(response.crossingById.latestStatus.id).toEqual(newStatusId);
    expect(response.crossingById.latestStatus.notes).toMatchSnapshot();
  });
}

describe('As a super admin', async () => {
  



  describe('when updating the status of a crossing', () => {
    describe('To OPEN', () => {
      describe('with no REASON or DURATION', async () => {
        shouldWork(superAdminEmail, superAdminPassword, 1, 3, 'OPEN with no REASON or DURATION');
      });

      // describe('with REASON', () => {
      //   it('should fail to update the status', () => {

      //   });
      // });

      // describe('with DURATION', () => {
      //   it('should fail to update the status', () => {

      //   });
      // });

      // describe('with REASON and DURATION', () => {
      //   it('should fail to update the status', () => {

      //   });
      // });
    });
  });



  // describe('To CLOSED', () => {
  //   describe('with no REASON or DURATION', () => {
  //     it('should fail to update the status', () => {

  //     });
  //   });

  //   describe('with REASON', () => {
  //     it('should fail to update the status', () => {

  //     });
  //   });

  //   describe('with DURATION', () => {
  //     it('should fail to update the status', () => {

  //     });
  //   });

  //   describe('with REASON and DURATION', () => {
  //     it('should fail to update the status', () => {

  //     });
  //   });
  // });


});
