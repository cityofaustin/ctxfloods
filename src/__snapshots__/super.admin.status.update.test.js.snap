// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As a super admin when updating the status of a crossing To OPEN with no REASON or DURATION should update the status 1`] = `"OPEN with no REASON or DURATION"`;

exports[`As a super admin when updating the status of a crossing To OPEN with no REASON or DURATION the new status should show up in the DB 1`] = `"OPEN with no REASON or DURATION"`;
